### ISSUE TITLE
generate_quantities can fail with simplex

### ISSUE THREAD

**Issue by avehtari** [stan-dev/cmdstanr#420]
Running generated quantities block that works fine with `sampling()` method can fail with `generate_quantities()` method. This is caused as by default `sig_figs=6` and thus the mapping to rounded stored values and back can produce wrong result. This is not a bug, but the error message is not helpful. Possible actions
- allow skipping the argument check and instead of an error give a warning
- or help the user by mentioning in the error message that the problem can be due to low value of `sig_figs`
The fixes might actually go to cmdstan, but as my example below uses cmdstanr I'm submitting the issue here.

Reproducible example (with high probability).
Stan code
```
data { vector[2] mu; }
parameters { simplex[2] x; }
model { x ~ normal(mu, 0.1); }
generated quantities { real y = sum(x); }
```
Sampling goes fine
```
datat <- list(mu=array(c(0,1)));
fitt <- modelt$sample(chains=1, refresh=0, data=datat)
```
Generating quantities
```
gq <- modelt$generate_quantities(fitt, data=datat)
```
fails with an error
```
Chain 1 Exception: stan::math::simplex_free: Simplex variable is not a valid simplex. sum(Simplex variable) = 0.9999998, but should be 1 (in '/tmp/RtmpqJOJQA/model-330c1654e1f.stan', line 5, column 2 to column 15)
Warning: Chain 1 finished unexpectedly!
```
The same example works with `datat <- list(mu=array(c(0.3,0.6)));`
If it would happen that the user would have a very time consuming sampling result, it's possible to work around the constraint by creating another model code with unrestricted parameter type.
```
data { vector[2] mu; }
parameters { vector[2] x; }
generated quantities { real y = sum(x); }
```

**Comment by jgabry** [comment]
Thanks Aki. I like the idea of hinting to the user in the error message that the problem could be due to a small value of `sig_figs`. 

@rok-cesnovar I think it would be preferable to do this in CmdStan itself since this would be an issue for CmdStanPy as well, but would it be possible possible for the CmdStan generate quantities method to catch and modify exceptions from Stan math?

**Comment by rok-cesnovar** [comment]
> but would it be possible possible for the CmdStan generate quantities method to catch and modify exceptions from Stan math

Right now no and I am not really sure how big of a refactor that would be

The cleanest solution for these cases would definitely be to finally move to a binary output format. That would fix a bunch of these precision issues.

**Comment by ssp3nc3r** [comment]
It isn't just with `simplex` types. I just ran into this issue again with the new `sum_to_zero_vector`. My fix in the past and for this has been changing the type in the new program to a vector, but it seems instead what whatever significant figures are chosen for output, it should round in a way that adheres to the type constraints, no? Or getting the binary format, which should also greatly speed up things to boot!

**Comment by jgabry** [comment]
> It isn't just with `simplex` types. I just ran into this issue again with the new `sum_to_zero_vector`.

Yeah that makes sense given the constraint imposed by `sum_to_zero_vector`. 

>  it seems instead what whatever significant figures are chosen for output, it should round in a way that adheres to the type constraints, no? 

That would be nice. I think something like that should probably go in CmdStan itself, if doable.

**Comment by avehtari** [comment]
The quickest solutions would be either
- increase the default `sig_figs` to 8 in CmdStan
- change the constraint check to use only 6 digits in Stan

It is unfortunate that there is a mismatch between these two

**Comment by jgabry** [comment]
Going to close this since I think it's now fixed by https://github.com/stan-dev/cmdstan/pull/1308, which will be included in the upcoming release. If it's still an issue we can reopen.
